### DISCUSSION TITLE
Handling executeChoice Return Values Across Multiple Macros in QuickAdd

### DISCUSSION THREAD

**Discussion by lucasbcgeo** [chhoumann/quickadd#679] [Q&A]
Hello everyone,

I am developing a multi-script system in Obsidian using the QuickAdd plugin to automate the creation and management of notes based on user-selected themes. The system involves three main scripts that determine the name/path, frontmatter, and content/backmatter of notes.

Issue:
While the scripts are executing and the data is being correctly generated and logged, the expected results are not being returned to the main script. As shown in the console output, the auxiliary script successfully generates the required data, but it fails to return these data back to the main script, leading to undefined results.

Example of Issue:

The auxiliary script nameAndPath.js is supposed to return an object with properties like path, name, and success.
The main script calls this auxiliary script using quickAddApi.executeChoice() and expects to receive the object.
Despite the auxiliary script logging the correct return object, the main script receives undefined.

Request:
I would appreciate any insights or suggestions on why the results are not being returned as expected and how to ensure that the auxiliary script's output is correctly captured by the main script. Any advice or examples of handling similar situations in Obsidian or with the QuickAdd plugin would be extremely helpful!

Thank you for your help!

**Comment by chhoumann** [comment]
Coming back to this one late - sorry about that.

What you're seeing is by design, not a bug in your scripts: `executeChoice()` doesn't return anything, and the variables set inside the executed choice are cleared when it finishes. There is no return channel, so the `undefined` is expected - your auxiliary script's return object is simply never consumed.

Two ways to structure this that work today:

**Recipe A (recommended): one macro, three user-script steps** (nameAndPath.js, frontmatter.js, content.js in order). All steps in a single macro share `params.variables` - that map is the data bus. The first script writes `params.variables.path = ...` and `params.variables.name = ...`; later steps read `params.variables.path`. Note that a step's ordinary `return` value isn't consumed either; the variables map is the only channel between steps.

**Recipe B: don't route pure logic through choices.** If the auxiliary scripts are really helper functions rather than full QuickAdd choices, keep them as plain modules and `require()` them from the main script by absolute path (resolved from `app.vault.adapter.getBasePath()`), then call them as normal functions with normal return values.

`executeChoice` is for triggering complete choices, fire-and-forget style, and returning values from it isn't planned. The shared-variables model is documented here: https://quickadd.obsidian.guide/docs/Choices/MacroChoice
